### PR TITLE
RSDK-3697: break up rtk

### PR DESCRIPTION
**Pull request by susmitaSanyal**
This PR breaks up the gpsrtk model into gpsrtkserial and gpsrtki2c. Changes include: 
- New config and attributes for gps-serial and gps-i2c. 
- Adjustment to the existing gpsrtk attribute
- Moving supporting functions to utils 

Note: No tests were added, this will be handled in another PR. Reconfigure will also be added later.

### CODE REVIEW

File: components/movementsensor/gpsrtkserial/gps_rtk_serial.go [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+	serialStr = "serial"
+	ntripStr  = "ntrip"
+)
+
+// Config is used for converting NMEA MovementSensor with RTK capabilities config attributes.
+type Config struct {
+	NmeaDataSource           string `json:"nmea_data_source"`
+	SerialPath               string `json:"serial_path"`
+	SerialBaudRate           int    `json:"serial_baud_rate,omitempty"`
+	SerialCorrectionPath     string `json:"serial_correction_path,omitempty"`
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkserial/gps_rtk_serial.go:46]: might be wrong but i think serial correction path is no longer needed now that this driver only does ntrip, I think the `serialcorrectionpath` and `serialcorrectionbaudrate`

File: components/movementsensor/utils.go [hunk 2]
```diff
@@ -178,3 +178,22 @@ func (lp *LastPosition) IsZeroPosition(p *geo.Point) bool {
 func (lp *LastPosition) IsPositionNaN(p *geo.Point) bool {
 	return math.IsNaN(p.Lng()) && math.IsNaN(p.Lat())
 }
+
+// AddChk adds PMTK checksums to commands by XORing the bytes together.
+func AddChk(data []byte) []byte {
```
Comment (oliviamiller) [inline components/movementsensor/utils.go:183]: thanks for adding these to utils!!

**Review by oliviamiller — CHANGES_REQUESTED** [review]

File: components/movementsensor/gpsrtkserial/gps_rtk_serial.go [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+)
+
+const (
+	serialStr = "serial"
+	ntripStr  = "ntrip"
+)
+
+// Config is used for converting NMEA MovementSensor with RTK capabilities config attributes.
+type Config struct {
+	NmeaDataSource           string `json:"nmea_data_source"`
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkserial/gps_rtk_serial.go:43]: Scope doc doesn't include nmea_data_source in i2c. besides we can go without it.

[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkserial/gps_rtk_serial.go:43]: Looking to remove this in serial, if possible

File: components/movementsensor/gpsrtk/gpsrtk.go [hunk 4]
```diff
@@ -55,7 +55,7 @@ type Config struct {
 
 // NtripConfig is used for converting attributes for a correction source.
 type NtripConfig struct {
-	NtripAddr            string `json:"ntrip_addr"`
+	NtripURL             string `json:"ntrip_addr"`
 	NtripConnectAttempts int    `json:"ntrip_connect_attempts,omitempty"`
 	NtripMountpoint      string `json:"ntrip_mountpoint,omitempty"`
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtk/gpsrtk.go:60]: I want to keep gpsrtk alive until I'm done adding reconfig and tests.

File: components/movementsensor/gpsrtki2c/gps_rtk_i2c.go [hunk 5]
```diff
@@ -0,0 +1,623 @@
+// Package gpsrtki2c implements a gps using serial connection
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtki2c/gps_rtk_i2c.go:1]: Good one, I'll add some description

File: components/movementsensor/gpsrtki2c/gps_rtk_i2c.go [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+			return g.err.Get()
+		}
+	}
+
+	errMsg := fmt.Sprintf("Can't connect to NTRIP caster after %d attempts", maxAttempts)
+	return errors.New(errMsg)
+}
+
+// GetStream attempts to connect to ntrip streak until successful connection or timeout.
+func (g *RTKI2C) GetStream(mountPoint string, maxAttempts int) error {
```
Comment (randhid) [inline components/movementsensor/gpsrtki2c/gps_rtk_i2c.go:224]: do non-api methods in this package need to be exported?

File: components/movementsensor/gpsrtki2c/gps_rtk_i2c.go [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+	resource.RegisterComponent(
+		movementsensor.API,
+		rtkmodel,
+		resource.Registration[movementsensor.MovementSensor, *Config]{
+			Constructor: newRTKI2C,
+		})
+}
+
+// RTKI2C is an nmea movementsensor model that can intake RTK correction data via I2C.
+type RTKI2C struct {
```
Comment (randhid) [inline components/movementsensor/gpsrtki2c/gps_rtk_i2c.go:93]: do the structs in this package need to be exported?

File: components/movementsensor/gpsrtki2c/gps_rtk_i2c.go [hunk 8]
```diff
@@ -0,0 +1,613 @@
+// Package gpsrtki2c implements a gps using serial connection
```
Comment (randhid) [inline components/movementsensor/gpsrtki2c/gps_rtk_i2c.go:1]: [nit] I think we like/prefer the filenames that have the package meat to match the package name, but I'm not too bothered by that. It's just easier for someone looking through our code base to search for using an exact matching package name.

File: components/movementsensor/register/register.go [hunk 9]
```diff
@@ -8,6 +8,8 @@ import (
 	_ "go.viam.com/rdk/components/movementsensor/fake"
 	_ "go.viam.com/rdk/components/movementsensor/gpsnmea"
 	_ "go.viam.com/rdk/components/movementsensor/gpsrtk"
```
Comment (randhid) [inline components/movementsensor/register/register.go:10]: You can de-register this model, but keep the rtk code.

File: components/movementsensor/gpsrtkserial/gpsrtkserial.go [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+		defer g.ntripMu.Unlock()
+		return spatialmath.NewZeroOrientation(), lastError
+	}
+	g.ntripMu.Unlock()
+
+	return g.nmeamovementsensor.Orientation(ctx, extra)
+}
+
+// ReadFix passthrough.
+func (g *rtkSerial) ReadFix(ctx context.Context) (int, error) {
```
Comment (randhid) [inline components/movementsensor/gpsrtkserial/gpsrtkserial.go:456]: Make non-API methods unexported where you can.

File: components/movementsensor/gpsrtk/gpsrtk.go [hunk 11]
```diff
@@ -192,8 +192,8 @@ func (cfg *SerialConfig) ValidateSerial(path string) error {
 
 // ValidateNtrip ensures all parts of the config are valid.
 func (cfg *NtripConfig) ValidateNtrip(path string) error {
-	if cfg.NtripAddr == "" {
-		return utils.NewConfigValidationFieldRequiredError(path, "ntrip_addr")
+	if cfg.NtripURL == "" {
+		return utils.NewConfigValidationFieldRequiredError(path, "ntrip_url")
```
Comment (randhid) [inline components/movementsensor/gpsrtk/gpsrtk.go:196]: I think you can get rid of the config, model, and init functions in this file, the creator and nmea struct can stayl

File: components/movementsensor/gpsrtkserial/gpsrtkserial.go [hunk 12]
```diff
@@ -0,0 +1,567 @@
...
+	}
+
+	g.logger.Debug("Connected to NTRIP caster")
+	g.ntripMu.Lock()
+	g.ntripClient.Client = c
+	g.ntripMu.Unlock()
+	return g.err.Get()
+}
+
+// getStream attempts to connect to ntrip streak until successful connection or timeout.
```
[the remark this answers is not in the document]
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkserial/gpsrtkserial.go:219]: I'll handle it in a separate PR if that's alright.

File: components/movementsensor/gpsrtk/gpsrtk.go (hunk 11 above)
Comment (randhid) [reply components/movementsensor/gpsrtk/gpsrtk.go:196]: Please remove the dead code in this next pr.

**Review by randhid — APPROVED** [review]

File: components/movementsensor/gpsrtk/gpsrtk.go (hunk 11 above)
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtk/gpsrtk.go:196]: Sure!

**Review by oliviamiller — APPROVED** [review]

**Comment by randhid** [comment]
To unblock Olivia could you make the parser and GPSdata struct public as well?

**Pull request merged**